### ISSUE TITLE
sharedstorage.php throws exception, making data inaccessible

### ISSUE THREAD

**Issue by LukasReschke** [owncloud/core#22976]
@danimo and @jospoortvliet are not able to access their data on their updated 9.0 instance unless https://github.com/owncloud/core/blob/c430f5ba53e02efb373667c8baf59f0209d3bfeb/lib/private/files/view.php#L1639-L1639 has been removed.

**If somebody encounters the exception one can remove https://github.com/owncloud/core/blob/c430f5ba53e02efb373667c8baf59f0209d3bfeb/lib/private/files/view.php#L1639-L1639 as a workaround until we have a proper fix for ownCloud 9.0.1**

The exception being:
```
atal   webdav  Exception: {"Message":"File with id \"3207194\" has not been found.","Exception":"OCP\\Files\\NotFoundException","Code":0,"Trace":"#0 \/srv\/www\/htdocs\/owncloud\/apps\/files_sharing\/lib\/sharedstorage.php(82): OC\\Files\\View->getPath(3207194)\n#1 \/srv\/www\/htdocs\/owncloud\/apps\/files_sharing\/lib\/sharedstorage.php(570): OC\\Files\\Storage\\Shared->init()\n#2 \/srv\/www\/htdocs\/owncloud\/lib\/private\/files\/storage\/wrapper\/wrapper.php(381): OC\\Files\\Storage\\Shared->getCache('', Object(OCA\\Files_Trashbin\\Storage))\n#3 \/srv\/www\/htdocs\/owncloud\/lib\/private\/files\/view.php(1302): OC\\Files\\Storage\\Wrapper\\Wrapper->getCache('')\n#4 \/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/connector\/sabre\/serverfactory.php(124): OC\\Files\\View->getFileInfo('\/jospoortvliet\/...')\n#5 [internal function]: OCA\\DAV\\Connector\\Sabre\\ServerFactory->OCA\\DAV\\Connector\\Sabre\\{closure}(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#6 \/srv\/www\/htdocs\/owncloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\n#7 \/srv\/www\/htdocs\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(446): Sabre\\Event\\EventEmitter->emit('beforeMethod', Array)\n#8 \/srv\/www\/htdocs\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(248): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#9 \/srv\/www\/htdocs\/owncloud\/apps\/dav\/appinfo\/v1\/webdav.php(55): Sabre\\DAV\\Server->exec()\n#10 \/srv\/www\/htdocs\/owncloud\/remote.php(138): require_once('\/srv\/www\/htdocs...')\n#11 {main}","File":"\/srv\/www\/htdocs\/owncloud\/lib\/private\/files\/view.php","Line":1639,"User":"jospoortvliet"}  2016-03-08T22:36:37+00:00
Error   no app in context   The folder is not available: File with id "3207194" has not been found.     2016-03-08T22:36:37+00:00
Fatal   webdav  Exception: {"Message":"File with id \"3207194\" has not been found.","Exception":"OCP\\Files\\NotFoundException","Code":0,"Trace":"#0 \/srv\/www\/htdocs\/owncloud\/apps\/files_sharing\/lib\/sharedstorage.php(82): OC\\Files\\View->getPath(3207194)\n#1 \/srv\/www\/htdocs\/owncloud\/apps\/files_sharing\/lib\/sharedstorage.php(570): OC\\Files\\Storage\\Shared->init()\n#2 \/srv\/www\/htdocs\/owncloud\/lib\/private\/files\/storage\/wrapper\/wrapper.php(381): OC\\Files\\Storage\\Shared->getCache('', Object(OCA\\Files_Trashbin\\Storage))\n#3 \/srv\/www\/htdocs\/owncloud\/lib\/private\/files\/view.php(1302): OC\\Files\\Storage\\Wrapper\\Wrapper->getCache('')\n#4 \/srv\/www\/htdocs\/owncloud\/apps\/dav\/lib\/connector\/sabre\/serverfactory.php(124): OC\\Files\\View->getFileInfo('\/jospoortvliet\/...')\n#5 [internal function]: OCA\\DAV\\Connector\\Sabre\\ServerFactory->OCA\\DAV\\Connector\\Sabre\\{closure}(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#6 \/srv\/www\/htdocs\/owncloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\n#7 \/srv\/www\/htdocs\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(446): Sabre\\Event\\EventEmitter->emit('beforeMethod', Array)\n#8 \/srv\/www\/htdocs\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(248): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#9 \/srv\/www\/htdocs\/owncloud\/apps\/dav\/appinfo\/v1\/webdav.php(55): Sabre\\DAV\\Server->exec()\n#10 \/srv\/www\/htdocs\/owncloud\/remote.php(138): require_once('\/srv\/www\/htdocs...')\n#11 {main}","File":"\/srv\/www\/htdocs\/owncloud\/lib\/private\/files\/view.php","Line":1639,"User":"jospoortvliet"}  2016-03-08T22:36:13+00:00
```

My guess is that the recent cleanup of the shared cache is doing something wrong here. This has been added on January 22 https://github.com/owncloud/core/commit/1369f2c6840e6bbac50526236771cb730a74f975

Somehow the owner view does not seem to have access to said file. See also https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/sharedstorage.php#L82

cc @icewind1991

**Comment by LukasReschke** [comment]
We applied the following patch there for more logging output, this is from @danimo's server:

```diff
From a75b2c13b251fcd537b378efa80a9b87eb62c017 Mon Sep 17 00:00:00 2001
From: Lukas Reschke <lukas@owncloud.com>
Date: Wed, 9 Mar 2016 00:15:30 +0100
Subject: [PATCH]

---
 apps/files_sharing/lib/sharedstorage.php | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/apps/files_sharing/lib/sharedstorage.php b/apps/files_sharing/lib/sharedstorage.php
index 6998f94..8552f6a 100644
--- a/apps/files_sharing/lib/sharedstorage.php
+++ b/apps/files_sharing/lib/sharedstorage.php
@@ -79,7 +79,12 @@ private function init() {
 		}
 		$this->initialized = true;
 		Filesystem::initMountPoints($this->share['uid_owner']);
-		$sourcePath = $this->ownerView->getPath($this->share['file_source']);
+		try {
+			$sourcePath = $this->ownerView->getPath($this->share['file_source']);
+		} catch (\Exception $e) {
+			var_dump($this->share);
+			exit();
+		}
 		list($this->sourceStorage, $sourceInternalPath) = $this->ownerView->resolvePath($sourcePath);
 		$this->sourceRootInfo = $this->sourceStorage->getCache()->get($sourceInternalPath);
 	}
```

The output here is: 

```

array(21) {
  ["id"]=>
  int(65)
  ["item_type"]=>
  string(6) "folder"
  ["item_source"]=>
  string(5) "25173"
  ["item_target"]=>
  string(6) "/25173"
  ["parent"]=>
  NULL
  ["share_type"]=>
  int(0)
  ["share_with"]=>
  string(5) "horst"
  ["uid_owner"]=>
  string(35) "3171d281-b73c11e2-be138443-e451810b"
  ["file_source"]=>
  int(25173)
  ["path"]=>
  string(19) "files/shareforhorst"
  ["file_target"]=>
  string(14) "/shareforhorst"
  ["permissions"]=>
  int(17)
  ["stime"]=>
  int(1413990064)
  ["expiration"]=>
  NULL
  ["token"]=>
  NULL
  ["storage"]=>
  int(5)
  ["mail_send"]=>
  string(1) "0"
  ["storage_id"]=>
  string(12) "home::mgoetz"
  ["file_parent"]=>
  int(2579)
  ["share_with_displayname"]=>
  string(12) "Horst Horsti"
  ["displayname_owner"]=>
  string(13) "Goetz, Markus"
}
```

**Comment by LukasReschke** [comment]
```
MariaDB [owncloud]> select * from oc_filecache where fileid="25173";
+--------+---------+---------------------+----------------------------------+--------+---------------+----------+----------+---------+------------+-----------+---------------+------------------+---------------+-------------+----------+
| fileid | storage | path                | path_hash                        | parent | name          | mimetype | mimepart | size    | mtime      | encrypted | etag          | unencrypted_size | storage_mtime | permissions | checksum |
+--------+---------+---------------------+----------------------------------+--------+---------------+----------+----------+---------+------------+-----------+---------------+------------------+---------------+-------------+----------+
|  25173 |       5 | files/shareforhorst | 8bd4205c61fb1b28c5519a540ce1346f |   2579 | shareforhorst |        1 |        2 | 8012236 | 1432037096 |         0 | 555b26e941b18 |                0 |    1430225956 |          31 | NULL     |
+--------+---------+---------------------+----------------------------------+--------+---------------+----------+----------+---------+------------+-----------+---------------+------------------+---------------+-------------+----------+
```

```
MariaDB [owncloud]> select * from oc_storages where numeric_id=5;
+--------------+------------+-----------+--------------+
| id           | numeric_id | available | last_checked |
+--------------+------------+-----------+--------------+
| home::mgoetz |          5 |         1 |         NULL |
+--------------+------------+-----------+--------------+
```

**Comment by nickvergessen** [comment]
Can we get a stacktrace from the exception please?

**Comment by LukasReschke** [comment]
@icewind1991 Any idea? THX. 

I was able to reproduce this locally by clicking around like a crazy but can't reproduce it anymore after trying to find out the steps :cry:

**Comment by danimo** [comment]
Is there any way I can help?

**Comment by icewind1991** [comment]
>   ["uid_owner"]=>
  string(35) "3171d281-b73c11e2-be138443-e451810b"


>   ["storage_id"]=>
  string(12) "home::mgoetz"

I would expect these to match

**Comment by icewind1991** [comment]
are you using an ldap setup?

**Comment by icewind1991** [comment]
https://github.com/owncloud/core/pull/23039 should at least prevent owncloud from breaking if there is an invalid share

**Comment by LukasReschke** [comment]
Another report in the forum can be found at https://forum.owncloud.org/viewtopic.php?f=38&t=34198

**Comment by jospoortvliet** [comment]
I sure don't use LDAP, don't even have it enabled.

#23039 works for me.

**Comment by oparoz** [comment]
#23039 does not help in the case of #23057 23057

**Comment by oparoz** [comment]
wait, wait, wrong browser window...

**Comment by oparoz** [comment]
Yep, same thing. I can navigate through the broken share.

**Comment by PVince81** [comment]
@icewind1991 @LukasReschke any progress on the root issue ?

**Comment by PVince81** [comment]
Can we simulate what happens with non-matching LDAP users as per https://github.com/owncloud/core/issues/22976#issuecomment-194297795 ? @icewind1991 @blizzz

**Comment by blizzz** [comment]
@danimo what's the username of Markus?

**Comment by blizzz** [comment]
At this point of time I don't know how the mismatching storage + username are created. If I change the storage name manually, a new one will be created with the username and the dir contents are ("re")canned. Resulting in duplicated entries in the filecache.

@danimo does this return anything? ``SELECT * from oc_storages WHERE id = "home::3171d281-b73c11e2-be138443-e451810b"``

**Comment by blizzz** [comment]
When an LDAP user is looked up (Users page, sharing dialog) it mapped and also the storage entry is created. There does not seem to be room between when a user is mapped and the entry is created to change the LDAP config in between.

**Comment by PVince81** [comment]
@blizzz maybe it happened through older bugs from older versions and got carried on with updates ?
Let's see what @danimo finds out.

Then the question is, how would the sharing code react with shares with mismatched user names ? @rullzer

**Comment by PVince81** [comment]
@blizzz assigning to you now until the needed info is gathered

**Comment by blizzz** [comment]
In @danimo`es case these are two different users (one LDAP, mgoetz local). They share the same user folder, but of course have own storage entries. The problematic entry belonged to the local user. 

That means that there is no issue with different usernames and storage ids.

**Comment by blizzz** [comment]
Anyway, this common usage of the same user folder probably lead to that effect that the LDAP-Markus was sharing the file, which was connected to the Local-Markus's storage, I assume. Does it sound likely @icewind1991?

**Comment by blizzz** [comment]
Anyway, this is probably just a funny accident, since @jospoortvliet experiences the stuff with only local users.

**Comment by blizzz** [comment]
A current backtrace from @jospoortvliet, not much differing from the one in the description though.

```
Exception: {"Exception":"OCP\Files\NotFoundException","Message":"File with id \"3207194\" has not been found.","Code":0,"Trace":"
#0 /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php(90): OC\Files\View->getPath(3207194)
#1 /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php(581): OC\Files\Storage\Shared->init()
#2 /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php(381): OC\Files\Storage\Shared->getCache('', Object(OCA\Files_Trashbin\Storage))
#3 /srv/www/htdocs/owncloud/lib/private/files/view.php(1365): OC\Files\Storage\Wrapper\Wrapper->getCache('')
#4 /srv/www/htdocs/owncloud/lib/private/files/node/folder.php(80): OC\Files\View->getDirectoryContent('/jospoortvliet/...')
#5 /srv/www/htdocs/owncloud/apps/files_sharing/api/share20ocs.php(415): OC\Files\Node\Folder->getDirectoryListing()
#6 /srv/www/htdocs/owncloud/apps/files_sharing/api/share20ocs.php(470): OCA\Files_Sharing\API\Share20OCS->getSharesInDir(Object(OC\Files\Node\Folder))
#7 /srv/www/htdocs/owncloud/apps/files_sharing/api/ocssharewrapper.php(40): OCA\Files_Sharing\API\Share20OCS->getShares()
#8 /srv/www/htdocs/owncloud/lib/private/api.php(175): OCA\Files_Sharing\API\OCSShareWrapper->getAllShares(Array)
#9 [internal function]: OC_API::call(Array)
#10 /srv/www/htdocs/owncloud/lib/private/route/router.php(273): call_user_func(Array, Array)
#11 /srv/www/htdocs/owncloud/ocs/v1.php(52): OC\Route\Router->match('/ocs/apps/files...')
#12 /srv/www/htdocs/owncloud/ocs/v2.php(22): require_once('/srv/www/htdocs...')
#13 {main}","File":"/srv/www/htdocs/owncloud/lib/private/files/view.php","Line":1639}
```

We checked the shares of this file (a folder). There were two entries, a share between two local users and a link share. Additionally this file was residing on an external storage.

The folder was accessible by mount between those two users, which make the share in the database look strange. When looks at "Shared with me" in the Files sidebar, it does not appear.

He also tried to access the link share (by copying out the token). The result was not only a File not Found, but also a Redirect Loop to the same location.

In the Files sidebar the was no share information on that folder.

The storage of that file is prefixed "local::" and points to the path of the mount. 

Checking again the filecache table on that folder returned 6 entries with the same file name. Two of them related to the data directory storage, two to the user storages, and two to "local::" mountpoints.

Here it caught Jos's attention because the storage that is connected to the Exception is a path to the location of a former external storage. So it appears that this is an orphaned entry somehow lingering around.

**Comment by blizzz** [comment]
In danimo's case a folder we looked at also belongs to an orphaned storage (4). There is no user connected to its username.

I believe we can say this somehow relates to shares on orphaned storages.

**Comment by PVince81** [comment]
> In danimo's case a folder we looked at also belongs to an orphaned storage (4). There is no user connected to its username.

> I believe we can say this somehow relates to shares on orphaned storages.


@rullzer @blizzz can we simulate this situation locally and see how the sharing code behaves ? Ideally such error should be ignored and nothing should break.

**Comment by blizzz** [comment]
1. Have two users
2. Configure an external storage, local type (/tmp/foo) and enable sharing
3. As user A, create a folder in this mount and share it to user B. Also create a file in that folder.
4. As user B, log in, browse around, download that file
5. Outside of ownCloud create a symlink to /tmp/foo as /tmp/foobar
6. Change the external storage path from /tmp/foo to /tmp/foobar
7. Profit. local::/tmp/foo and local::/tmp/foobar both exists as storage. 
8. As user B, refresh files view.
9. Check owncloud Log. Exception, as posted earlier in this thread, is thrown.
10. Outside of ownCloud create a symlink to /tmp/foo as /tmp/barfoo
11. Remove the current external mount (/tmp/foobar) and create a new one (/tmp/barfoo)
12. Check oc_storages table: now the old storage local::/tmp/foobar was removed properly. A new one was created. Very old local::/tmp/foo still exists.

→ oc_storages is not cleaned up if an external storage is edited and the path changed

So far at least about @jospoortvliet's cases.

In @danimo`s, external storage did not play in, but I have an idea about it as well.

**Comment by PVince81** [comment]
@blizzz that's clever, using local ext storage to simulate orphaned entries! Nice one!

So if I understand properly, the share is pointing at an obsolete "oc_storages" entry and since that one isn't mounted (or is it?), it fails.

**Comment by PVince81** [comment]
@icewind1991 is there a way to make this more robust ?

@blizzz let us know when your second idea for @danimo's env pops up :smile:

**Comment by blizzz** [comment]
@PVince81 one idea was that storages are not removed on user deletion. But I could not confirm it about local users. The other one is that there was such an LDAP user A, who had the home folder in common with another one, and was eventually deleted on LDAP, but not deleted in ownCloud. Waiting on feedback from @danimo about it.

If true, this could be a problem in general if any user  is temporarily not available. While this is an edge case, it could make detecting orphaned storages more complicated.

**Comment by blizzz** [comment]
I got the confirmation from @danimo that the orphaned storage belonged to a deleted local user. Apparently some time somewhere cleanup was not that complete when deleting users. Possibly it's also from the old days when we did no cleanup at all when a user was deleted. 

So, what we basically need is a repair step or maintenance tool that goes through the existing storages and figure out which ones are orphaned. ``home:: storages`` by checking whether the user exists, ``local::storages`` by checking whether the mountpoint exists. If an orphaned storaged was found, all possible remainings in the database need to be purged (e.g. filecache, shares, preferences…).

**Comment by PVince81** [comment]
@blizzz let's make a separate ticket for this. The critical part "making data inaccessible" was solved through @icewind1991's fix, so I think the severity for the repair step is less than critical.

Would be good to find out how OC 8.2.3 behaves in that regard. I suspect that the storage API changes may have produced real issues while the old code might have just ignored it.

**Comment by PVince81** [comment]
Also to note: such repair step cannot be written pre-OC 9 because there was no way to resolve a user id from an oc_storages entry. Since OC 9 there is the table "oc_mounts" that contains such information, so it could be used for that, given that the table is left as is after user deletion...

**Comment by blizzz** [comment]
As discussed with @PVince81, the criticalness here was solved with https://github.com/owncloud/core/pull/23039, for dealing with the root problem please follow https://github.com/owncloud/core/issues/23364
